feat(taproot): add full taproot support

mariocynicys opened the pull request:
This PR adds full taproot support (address parsing, address generation, withdrawls, signing, trezor, etc...).

For now, this adds support for native only (no wasm) until https://github.com/KomodoPlatform/komodo-defi-framework/pull/2623 is merged. This is due to the same issue related to multiple secp dependencies co-existing at the same time.

Tests to be added.

Closes https://github.com/KomodoPlatform/komodo-defi-framework/pull/2625 (since it didn't get merged early enough)

mariocynicys committed:
remove signature_version() method

we don't need to gaurd a segwit address, segwit addresses will be properly detected inside with_key_pair::sign_tx() using the prev_script

mariocynicys committed:
taproot starter

filler code for signing and recognition of taproot addresses etc...

mariocynicys committed:
add version field for taproot preparation

his commit refactors the UTXO address handling system to support versioned segwit addresses, laying the groundwork for taproot (segwit v1) implementation:

- **AddressFormat refactor**: Convert `AddressFormat::Segwit` to `AddressFormat::Segwit { version: u8 }` to support multiple segwit versions
- **Method updates**: Replace `is_segwit()` calls with `is_segwit_v0()` throughout the codebase to maintain existing segwit v0 behavior
- **Bech32m support**: Add bech32m encoding support for segwit v1+ addresses in SegwitAddress display formatting
- **Address construction**: Update SegwitAddress constructor to accept version parameter
- **Script type mapping**: Add P2TR script type detection and mapping for taproot addresses

- Segwit v0 addresses continue to use bech32 encoding
- Segwit v1+ addresses use bech32m encoding as per BIP-350
- Added version validation and proper script type detection
- Maintains backward compatibility with existing segwit v0 functionality
- Includes FIXME comments marking areas requiring taproot-specific implementation

- Taproot signing logic in transaction generation
- P2TR output script handling in spending operations
- Hardware wallet support for taproot addresses
- Complete taproot validation and witness handling

mariocynicys committed:
add taproot signing support

but not yet for wasm!

since taproot signing uses rust-bitcoin (to no re-invent the wheel), adding rust-bitcoin dependency on wasm will break linking just
like what happened in https://github.com/KomodoPlatform/komodo-defi-framework/pull/2566

since https://github.com/KomodoPlatform/komodo-defi-framework/pull/262 already fixes this issue by converging all versions
of secp (and thus rust-bitcoin) and enabling them on wasm environment, let's defer wasm support till that PR is merged

mariocynicys committed:
add variant AddressHashEnum::TweakedXOnlyPubkey and fix a couple of fixmes

mariocynicys committed:
rename as_pkh_from_pk to using_pk

as it could be used now to create taproot addresses

mariocynicys committed:
add a check to not use bad build option with segwit version

mariocynicys committed:
remove some fixme and refine another

mariocynicys committed:
side-quest: ease fork_id mambo jambo

don't involve fork_id logic in Sighash. also remove the abstracting function signature_hash_fork_id and replace it by signature_hash_witness0

mariocynicys committed:
side-quest: ease fork_id mambo jambo cont.

based on our coins config, there exists no fork_id chain that doens't have the fork_id bit set. thus we don't really need to handle a fork_id chain that doesn't have the fork_id bit set and is required to be signed as signature_hash_original (huh?) cause this DOESN'T EXIST

mariocynicys committed:
disallow parsing witness_v0 signature_version

since it's already detectable from within KDF. the variant is kept though to be used internally as a flag on how to sign the (witness) transaction

mariocynicys committed:
remove leftover SignatureVersion from Sighash

mariocynicys committed:
don't use witnesscripthash for taproot signing

that's a stray witnessscripthash that i forgot to change

mariocynicys committed:
resolve the fixme related to tweaking user-provided pubkeys

mariocynicys committed:
update trezor protos to add taproot inputs and outputs

this resolves a couple of fixmes as well

mariocynicys committed:
support full taproot spending (incl. trezor)

mariocynicys committed:
resolve a fixme by not assuming that the segwit address version is correct

error if it's not correct

mariocynicys committed:
bulk rename AddressHashEnum to LockingDestination

this commit doesn't bring any logical change, don't worry ;D

mariocynicys committed:
fix: don't use witness script hash for taproot

mariocynicys committed:
fix last bits to get taproot enabling working

haven't tested withdrawing yet

mariocynicys committed:
adapt the taproot test to allow withdrawing to taproot address

mariocynicys committed:
add taproot address parsing tests

brough from here: https://github.com/KomodoPlatform/komodo-defi-framework/pull/2625

mariocynicys committed:
add a fixme

onur-ozkan commented:
Fmt & linting pipelines are failing.

onur-ozkan commented on mm2src/mm2_bitcoin/keys/src/segwitaddress.rs:
```diff
@@ -76,24 +79,35 @@ pub struct SegwitAddress {
+            1 => match self.program.len() {
+                32 => Some(SegwitAddrType::P2tr),
+                _ => None,
+            },
```
See https://github.com/KomodoPlatform/komodo-defi-framework/pull/2625#discussion_r2412613920.

mariocynicys replied:
[6689d5f](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2630/commits/6689d5f090b75d107f457cec8f4ad49ff4f77f84)

onur-ozkan commented on mm2src/mm2_bitcoin/keys/src/address.rs:
```diff
@@ -281,48 +306,68 @@ impl Address {
-            (AddressScriptType::P2WSH, AddressHashEnum::default_witness_script_hash())
-        } else {
-            return Err("Expect either 20 or 32 bytes long hash".into());
+        let (script_type, mut locking_destination) = match (address.version_as_u8(), address.program.len()) {
+            (0, 20) => (AddressScriptType::P2WPKH, LockingDestination::default_address_hash()),
+            (0, 32) => (
+                AddressScriptType::P2WSH,
+                LockingDestination::default_witness_script_hash(),
+            ),
+            (0, _) => return Err("Expect either 20 or 32 bytes long hash for witness v0 address".into()),
+            (1, 32) => (
+                AddressScriptType::P2TR,
+                LockingDestination::default_tweaked_xonly_pubkey(),
+            ),
+            (1, _) => return Err("Expect 32 bytes long public key for witness v1 address".into()),
+            (v, _) => return Err(format!("Unsupported segwit version: {v}")),
+        };
```
See https://github.com/KomodoPlatform/komodo-defi-framework/pull/2625#discussion_r2412619430.

mariocynicys replied:
i moved all the wild cards to the end (namely, `(0, _)` to the end).
do u mean something like having a wild card at the end for all the wild cards and match which error message we wanna return in it?

something like:
```rust
match (version, length) {
  (0, 20) => ...,
  (0, 32) => ...,
  (1, 32) => ...,
  (v, _l) => match v {
               0 => Err("Expect either 20 or 32 bytes long hash for witness v0 address"),
               1 => Err("Expect 32 bytes long public key for witness v1 address"),
               v => Err(format!("Unsupported segwit version: {v}")),
             }
}
```

in such matching, as long as the only wild card arm (i.e. `(v, _l)` arm) is at the end, we won't face any bug.

onur-ozkan commented on mm2src/mm2_bitcoin/keys/src/address.rs:
```diff
@@ -281,48 +306,68 @@ impl Address {
-        match &self.hrp {
-            Some(hrp) => Ok(SegwitAddress::new(&self.hash, hrp.to_string())),
-            None => Err("hrp must be provided for segwit address".into()),
+        match (&self.hrp, &self.addr_format) {
+            (Some(hrp), AddressFormat::Segwit { version }) => {
+                Ok(SegwitAddress::new(&self.locking_destination, hrp.to_string(), *version)
+                    .map_err(|e| e.to_string())?)
+            },
+            _ => Err("hrp and segwit version must be provided for segwit address".into()),
```
It's better to break this logic into pieces, like:

(dummy code):

```rs
let Some(hrp) = self.hrp else {
   return Err("hrp must be provided for segwit address");
};

if let AddresssFormat::Segwit ... = self.addr_format {
   ...
} else {
   Err("Invalid address format, expected Segwit, found {}", self.addr_format)
}
```

This will be much easier to debug in case of `to_segwitaddress` being failed.

mariocynicys replied:
[6689d5f](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2630/commits/6689d5f090b75d107f457cec8f4ad49ff4f77f84)

mariocynicys committed:
fix mis-written test

mariocynicys committed:
Merge remote-tracking branch 'origin/dev' into taproot-support

mariocynicys committed:
review(onur): clearer matching and error messages

mariocynicys committed:
Merge remote-tracking branch 'origin/dev' into taproot-support

mariocynicys committed:
fix post-merge compilation error

mariocynicys committed:
resolve cashaddress/legacy fixme

change the conidition so to allow cashaddress and legacy address types interchanging between the conf and request

mariocynicys committed:
resolve a fixme regaring runtime panic

leaving the expect() as is. since we already have a lot of expects that enforce our invariants.
if we somehow fail to catch an invalid segwit version while constructing, then let's just panic here since this is a bug

mariocynicys committed:
add a swap test for taproot coin

mariocynicys committed:
merge with origin/dev

AGENTS.md still needs to be updated since it has some references of old AddressHashEnum which is now replaced by LockingDestination
